src/credentials/DeviceAttestationConstructor.cpp

```diff
@@ -0,0 +1,173 @@
+ *    limitations under the License.
+ */
+#include "DeviceAttestationConstructor.h"
+
+#include <lib/core/CHIPTLV.h>
+#include <lib/support/CodeUtils.h>
+#include <lib/support/logging/CHIPLogging.h>
+
+#include <iostream>
+
+namespace chip {
+namespace Credentials {
+
+// context tag positions
+enum
+{
+    kCertificationDeclarationTagId = 1,
+    kAttestationNonceTagId         = 2,
+    kTimestampTagId                = 3,
+    kFirmwareInfoTagId             = 4,
+};
+
+CHIP_ERROR DeconstructAttestationElements(const ByteSpan & attestationElements, ByteSpan & certificationDeclaration,
+                                          ByteSpan & attestationNonce, uint32_t & timestamp, ByteSpan & firmwareInfo,
+                                          ByteSpan * vendorReservedArray, size_t vendorReservedArraySize, uint16_t & vendorId,
+                                          uint16_t & profileNum)
+{
+    bool certificationDeclarationExists = false;
+    bool attestationNonceExists         = false;
+    bool timestampExists                = false;
+    bool firmwareInfoExists             = false;
+    size_t vendorReservedIdx            = 0;
+    TLV::TLVReader tlvReader;
+    TLV::TLVType containerType = TLV::kTLVType_Structure;
+
+    tlvReader.Init(attestationElements.data(), static_cast<uint32_t>(attestationElements.size()));
+    ReturnErrorOnFailure(tlvReader.Next(containerType, TLV::AnonymousTag));
+    ReturnErrorOnFailure(tlvReader.EnterContainer(containerType));
+
+    CHIP_ERROR error = CHIP_NO_ERROR;
+    while ((error = tlvReader.Next()) == CHIP_NO_ERROR)
+    {
+        uint64_t tag = tlvReader.GetTag();
+
+        if (TLV::IsContextTag(tag))
+        {
+            switch (TLV::TagNumFromTag(tag))
+            {
+            case kCertificationDeclarationTagId:
+                VerifyOrReturnError(certificationDeclarationExists == false, CHIP_ERROR_IM_MALFORMED_COMMAND_DATA_ELEMENT);
+                ReturnErrorOnFailure(tlvReader.Get(certificationDeclaration));
+                certificationDeclarationExists = true;
+                break;
+            case kAttestationNonceTagId:
+                VerifyOrReturnError(attestationNonceExists == false, CHIP_ERROR_IM_MALFORMED_COMMAND_DATA_ELEMENT);
+                ReturnErrorOnFailure(tlvReader.Get(attestationNonce));
+                attestationNonceExists = true;
+                break;
+            case kTimestampTagId:
+                VerifyOrReturnError(timestampExists == false, CHIP_ERROR_IM_MALFORMED_COMMAND_DATA_ELEMENT);
+                ReturnErrorOnFailure(tlvReader.Get(timestamp));
+                timestampExists = true;
+                break;
+            case kFirmwareInfoTagId:
+                VerifyOrReturnError(firmwareInfoExists == false, CHIP_ERROR_IM_MALFORMED_COMMAND_DATA_ELEMENT);
+                ReturnErrorOnFailure(tlvReader.Get(firmwareInfo));
+                firmwareInfoExists = true;
+                break;
```

Comment:
The method here does not address the fact that all TLV elements *must* be in tag-order. A simple way to do it would be to keep track of "last context tag number", and start it at -1. In each test case, add what the expected previous tag should be.

Reply:
ok, didn't see that constraint.

Reply:
Current patch makes sure tags are in order.

Read A.2.4, should we have two while loops (instead of one big while loop) -- first for the context tags, second for profile tags.